src/charm.py

```diff
@@ -162,6 +180,97 @@ def __init__(self, *args):
             self.vault_pki.on.certificate_creation_request,
             self._on_vault_pki_certificate_creation_request,
         )
+        self.framework.observe(
+            self.vault_autounseal_requires.on.vault_autounseal_details_ready,
+            self._configure,
+        )
+        self.framework.observe(
+            self.vault_autounseal_provides.on.vault_autounseal_initialize,
+            self._on_vault_autounseal_initialize,
+        )
+        self.framework.observe(
+            self.vault_autounseal_provides.on.vault_autounseal_destroy,
+            self._on_vault_autounseal_destroy,
+        )
+        self.framework.observe(
+            self.vault_autounseal_requires.on.vault_autounseal_provider_removed,
+            self._configure,
+        )
+
+    def _on_vault_autounseal_destroy(self, event: VaultAutounsealDestroy):
+        if not self.unit.is_leader():
+            return
+
+        vault = self._get_active_vault_client()
+        if vault is None:
+            logger.warning("Vault is not active, cannot disable vault autounseal")
+            return
+        vault.destroy_autounseal_credentials(event.relation.id, AUTOUNSEAL_MOUNT_PATH)
+
+    def _on_vault_autounseal_initialize(self, event: VaultAutounsealInitialize):
+        self._generate_and_set_autounseal_credentials(event.relation)
+
+    def _generate_and_set_autounseal_credentials(self, relation: Relation) -> None:
+        if not self.unit.is_leader():
+            return
+        vault = self._get_active_vault_client()
+        if vault is None:
+            logger.warning("Vault is not active, cannot generate autounseal credentials")
+            return
+
+        vault.enable_secrets_engine(SecretsBackend.TRANSIT, AUTOUNSEAL_MOUNT_PATH)
+
+        # FIXME: We should probably forward the key name in the relation data,
+        # but for now we just infer it on the requirer side, so we ignore it
+        # here.
```

Comment:
Humm, why chose the infer approach? I'd suggest providing it. This will allow us to change the scheme in the future if needed

Reply:
Ooops, I just forgot to remove this comment. We don't infer it anymore.

Reply:
https://github.com/canonical/vault-k8s-operator/pull/365/commits/6168f1b1077f8aacf63e8cbe3bcab657d815e71e